src/main/java/com/tenx/ms/retail/client/rest/ClientController.java

```diff
@@ -0,0 +1,121 @@
+    public Client getClientById(@ApiParam(name = "clientId", value = "Client id") @PathVariable() Long clientId) {
+        return clientService.findClientById(clientId);
+    }
+
+    @ApiOperation(value = "Returns all clients")
+    @ApiResponses( value = {
+        @ApiResponse(code = 200, message = "Client list returned successfully"),
+        @ApiResponse(code = 500, message = "Internal server error")
+    })
+    @RequestMapping(method = RequestMethod.GET)
+    public List<Client> getClients(){
+        return clientService.findAllClients();
+    }
+
+    @ApiOperation(value = "Creates a new client")
+    @ApiResponses( value = {
+        @ApiResponse(code = 201, message = "Client successfully created"),
+        @ApiResponse(code = 400, message = "Invalid parameters"),
+        @ApiResponse(code = 412, message = "Validation failure"),
+        @ApiResponse(code = 500, message = "Internal server error")
+    })
+    @RequestMapping(method = RequestMethod.POST)
+    @ResponseStatus(HttpStatus.CREATED)
+    public ResourceCreated<Long> createClient(@ApiParam(name = "client", value = "Client data", required = true) @RequestBody Client client){
+        Client newClient = clientService.insertClient(client);
+        return new ResourceCreated<Long>(newClient.getClientId());
+    }
+
+    @ApiOperation(value = "Updates a client given an id")
+    @ApiResponses(value = {
+        @ApiResponse(code = 200, message = "Client successfully updated"),
+        @ApiResponse(code = 400, message = "Invalid parameters"),
+        @ApiResponse(code = 404, message = "Client can't be found by id"),
+        @ApiResponse(code = 412, message = "Validation failure"),
+        @ApiResponse(code = 500, message = "Internal server error")}
+    )
+    @RequestMapping(value = {"/{clientId:\\d+}"}, method = RequestMethod.PUT)
+    public Client updateClient(@ApiParam(name = "clientId", value = "Client id") @PathVariable() Long clientId,
+                             @ApiParam(name = "client", value = "Client data", required = true) @RequestBody Client client) {
+        return clientService.updateClient(clientId, client);
+    }
+
+    @ApiOperation(value = "Deletes a client given an id")
+    @ApiResponses(value = {
+        @ApiResponse(code = 204, message = "Client successfully deleted"),
+        @ApiResponse(code = 404, message = "Client can't be found by id"),
+        @ApiResponse(code = 500, message = "Internal server error")}
+    )
+    @RequestMapping(value = {"/{clientId:\\d+}"}, method = RequestMethod.DELETE)
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    public void deleteClient(@ApiParam(name = "clientId", value = "Client id") @PathVariable() Long clientId) {
+        clientService.deleteClient(clientId);
+    }
+
+    @ResponseStatus(value = HttpStatus.PRECONDITION_FAILED)
+    @ExceptionHandler(ConstraintViolationException.class)
+    protected void handleConstraintViolationException(ConstraintViolationException ex,
```

Comment:
These 3 exceptions seem to be database errors. I suggest not to wrap them as precondition failed unless you have specific reason. I think they should just be returned as 500 by default so these handleXXXException can be removed.